api/v1/observer/observer.proto

```diff
@@ -71,7 +71,10 @@ message GetFlowsRequest {
     // Number of flows that should be returned. Incompatible with `since/until`.
     uint64 number = 1;
 
-    // Unused 2
+    // Dump all flows. Incompatible with `since/until`, `follow` and any filter.
```

Comment:
Dump + follow is a legit use-case and i'd rather it not be hacked via `since 100000h + follow` if we can avoid it. Are there specific reasons for omitting follow?

Reply:
I'm missing a ton of context here, but generally, by the time a dump operation finishes, many new events/flows might've already occurred before a follow operation is initiated. This might require extra handling on the server side.

Again, no idea about Hubble's event delivery mechanisms, but a follow operation should always be initiated first, then the dump needs to take place, events need to be de-duplicated and pruned so the watcher doesn't observe out-of-order state transitions (eg. a SYN+ACK before a SYN event, ..). It might be worth omitting this functionality initially to keep the complexity down, and lift the restriction later.

Reply:
I'm sorry for the lack of context and explanation here. I have updated the commit message and PR description to explain the difference between `last==uint64_max` and `dump` and to explain why it's not possible to implement follow-mode for dump requests. It might actually be possible to implement follow-mode, eg by using 2 readers (one that reads backwards and one that follows) but this would be more complicated to implement.

Reply:
Yup! After reading the updated PR description, it makes sense. Simply put, this is basically a raw dump of the array backing the ring, and, as such, has no relation to the location of readers/consumers. Supporting a 'follow' in the same request wouldn't really make sense. :+1:

Reply:
After having given this more thoughts, I'm actually tempted to give the 2 readers approach a go. This would render the actual implementation useless and no API update would be required and would be overall more powerful. The main thing that worries me a bit with this approach is the memory overhead as events from both readers (ie the follow reader and the backward reader) need to be kept in-memory until the backward reader catches up with the writer (worst case being a request in follow-mode + all history without any filter). I'll mark this PR as draft for the time being.

Reply:
Yes, memory usage could be an issue, but this operation only copies pointers, so the overhead should be quite minimal. Wouldn't it be preferable to dump the backing array instead of actually following the buffer? Under extreme load, a dump operation might not terminate if the reader keeps getting behind, so it might need to be bounded by a certain maximum amount of records.

Reply:
> memory usage could be an issue, but this operation only copies pointers, so the overhead should be quite minimal.

Except that in a worst case scenario, more events are kept as copies than the actual ring buffer size (pointers to events that have already been overwritten in the ring buffer). I agree though than in most cases the overheard should be minimal.

> Wouldn't it be preferable to dump the backing array instead of actually following the buffer?

This would require exposing the backing array which is undesirable as this is ring buffer internals. Unless maybe a new `Dump` method is added to the ring struct.

> Under extreme load, a dump operation might not terminate if the reader keeps getting behind, so it might need to be bounded by a certain maximum amount of records.

This can't happen with the current dump implementation as the reader reads backwards, ie it starts from the most recently written position and rewinds until it meets with the writer which is inevitable at some point.

Reply:
> more events are kept as copies than the actual ring buffer size

Yup, but shouldn't matter that much since the runtime will hold onto them anyway until GC runs.

As for the remaining points, I just mean we should keep a dump operation as simple and linear as possible, as it's a debugging tool. Providing guarantees about message ordering might not be necessary.